ccl/backupccl: TestBackupRestoreSystemJobsProgress failed

[(ccl/backupccl).TestBackupRestoreSystemJobsProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2513569&tab=buildLog) on [master@3cd932db33c133bff6a298f25d2516ecce1a2f54](https://github.com/cockroachdb/cockroach/commits/3cd932db33c133bff6a298f25d2516ecce1a2f54):

```
=== RUN   TestBackupRestoreSystemJobsProgress
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress204890354
    test_log_scope.go:73: use -show-logs to present logs inline
    backup_test.go:1366: expected progress to be in range [0.25, 0.75] but got 0.200000
E201211 16:29:28.693480 102662 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  making txn commit explicit failed for "unnamed" meta={id=68ff1924 pri=0.06312649 epo=0 ts=1607704168.593859468,0 min=1607704168.593859468,0 seq=6} lock=true stat=STAGING rts=1607704168.593859468,0 wto=false max=1607704168.593859468,0 int=1 ifw=5: result is ambiguous (server shutdown)
E201211 16:29:28.730435 94231 jobs/registry.go:662  error processing claimed jobs: could query for claimed jobs: select-running/get-claimed-jobs: context canceled
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress204890354
--- FAIL: TestBackupRestoreSystemJobsProgress (62.37s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreSystemJobsProgress PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreSystemJobsProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/backupccl).TestBackupRestoreSystemJobsProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2514771&tab=buildLog) on [master@05a192a4269270659174249ed5238b505e540e40](https://github.com/cockroachdb/cockroach/commits/05a192a4269270659174249ed5238b505e540e40):

```
=== RUN   TestBackupRestoreSystemJobsProgress
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress892272278
    test_log_scope.go:73: use -show-logs to present logs inline
    backup_test.go:1366: expected progress to be in range [0.25, 0.75] but got 0.200000
E201211 23:17:00.193291 101953 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  making txn commit explicit failed for "unnamed" meta={id=4c0d00af pri=0.01438349 epo=0 ts=1607728620.106420566,0 min=1607728620.106420566,0 seq=6} lock=true stat=STAGING rts=1607728620.106420566,0 wto=false max=1607728620.106420566,0 int=1 ifw=5: result is ambiguous (server shutdown)
E201211 23:17:00.230017 93311 jobs/registry.go:662  error processing claimed jobs: could query for claimed jobs: select-running/get-claimed-jobs: context canceled
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress892272278
--- FAIL: TestBackupRestoreSystemJobsProgress (61.84s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreSystemJobsProgress PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreSystemJobsProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/backupccl).TestBackupRestoreSystemJobsProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2518258&tab=buildLog) on [master@dffc015a670422fce4609985862170c99d208aff](https://github.com/cockroachdb/cockroach/commits/dffc015a670422fce4609985862170c99d208aff):

```
=== RUN   TestBackupRestoreSystemJobsProgress
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress247491893
    test_log_scope.go:73: use -show-logs to present logs inline
    backup_test.go:1366: expected progress to be in range [0.25, 0.75] but got 0.166667
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress247491893
--- FAIL: TestBackupRestoreSystemJobsProgress (49.56s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreSystemJobsProgress PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreSystemJobsProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/backupccl).TestBackupRestoreSystemJobsProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2518426&tab=buildLog) on [master@0c59bbfdc515b248fc55853f2261535daf962eee](https://github.com/cockroachdb/cockroach/commits/0c59bbfdc515b248fc55853f2261535daf962eee):

```
=== RUN   TestBackupRestoreSystemJobsProgress
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress078020431
    test_log_scope.go:73: use -show-logs to present logs inline
    backup_test.go:1366: expected progress to be in range [0.25, 0.75] but got 0.200000
E201214 18:32:47.654825 292128 kv/kvclient/kvcoord/txn_interceptor_committer.go:449  making txn commit explicit failed for "unnamed" meta={id=fd39c940 pri=0.00445498 epo=0 ts=1607970767.619142196,0 min=1607970767.619142196,0 seq=6} lock=true stat=STAGING rts=1607970767.619142196,0 wto=false max=1607970767.619142196,0 int=1 ifw=5: result is ambiguous (server shutdown)
    panic.go:617: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress078020431
--- FAIL: TestBackupRestoreSystemJobsProgress (54.26s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreSystemJobsProgress PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreSystemJobsProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(ccl/backupccl).TestBackupRestoreSystemJobsProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2518316&tab=buildLog) on [master@ac4d8208863201dde49cab30288dc3765a152102](https://github.com/cockroachdb/cockroach/commits/ac4d8208863201dde49cab30288dc3765a152102):

```
=== RUN   TestBackupRestoreSystemJobsProgress
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress078851456
    test_log_scope.go:73: use -show-logs to present logs inline
    backup_test.go:1366: expected progress to be in range [0.25, 0.75] but got 0.200000
E201214 17:53:52.540208 91778 kv/kvserver/reports/reporter.go:142  [n1,replication-reporter] failed to generate replication reports: failed to save range status report: upsert-replication-stats: failed to send RPC: sending to all replicas failed; last error: node unavailable; try another peer
    panic.go:636: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestBackupRestoreSystemJobsProgress078851456
--- FAIL: TestBackupRestoreSystemJobsProgress (61.06s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestBackupRestoreSystemJobsProgress PKG=./pkg/ccl/backupccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestBackupRestoreSystemJobsProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/ccl/backupccl/backup_processor.go
+++ b/pkg/ccl/backupccl/backup_processor.go
@@ -18,13 +18,14 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowexec"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/storage/cloudimpl"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
-	hlc "github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
@@ -237,6 +238,13 @@ func runBackupProcessor(
 					return errors.Wrapf(pErr.GoError(), "exporting %s", span.span)
 				}
 				res := rawRes.(*roachpb.ExportResponse)
+
+				if backupKnobs, ok := flowCtx.TestingKnobs().BackupRestoreTestingKnobs.(*sql.BackupRestoreTestingKnobs); ok {
+					if backupKnobs.RunAfterExportingSpanEntry != nil {
+						backupKnobs.RunAfterExportingSpanEntry(ctx)
+					}
+				}
+
 				files := make([]BackupManifest_File, 0)
 				var prog execinfrapb.RemoteProducerMetadata_BulkProcessorProgress
 				progDetails := BackupManifest_Progress{}

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -1295,32 +1295,22 @@ func (ip inProgressState) latestJobID() (int64, error) {
 func checkInProgressBackupRestore(
 	t testing.TB, checkBackup inProgressChecker, checkRestore inProgressChecker,
 ) {
-	// To test incremental progress updates, we install a store response filter,
-	// which runs immediately before a KV command returns its response, in our
-	// test cluster. Whenever we see an Export or Import response, we do a
-	// blocking read on the allowResponse channel to give the test a chance to
-	// assert the progress of the job.
 	var allowResponse chan struct{}
 	params := base.TestClusterArgs{}
 	knobs := base.TestingKnobs{
-		DistSQL: &execinfra.TestingKnobs{BackupRestoreTestingKnobs: &sql.BackupRestoreTestingKnobs{RunAfterProcessingRestoreSpanEntry: func(ctx context.Context) {
-			<-allowResponse
-		}}},
-		Store: &kvserver.StoreTestingKnobs{
-			TestingResponseFilter: func(ctx context.Context, ba roachpb.BatchRequest, br *roachpb.BatchResponse) *roachpb.Error {
-				for _, ru := range br.Responses {
-					switch ru.GetInner().(type) {
-					case *roachpb.ExportResponse:
-						<-allowResponse
-					}
-				}
-				return nil
-			},
-		}}
+		DistSQL: &execinfra.TestingKnobs{
+			BackupRestoreTestingKnobs: &sql.BackupRestoreTestingKnobs{
+				RunAfterExportingSpanEntry: func(_ context.Context) {
+					<-allowResponse
+				},
+				RunAfterProcessingRestoreSpanEntry: func(_ context.Context) {
+					<-allowResponse
+				},
+			}},
+	}
 	params.ServerArgs.Knobs = knobs
 
 	const numAccounts = 1000
-	const totalExpectedResponses = backupRestoreDefaultRanges
 
 	ctx, _, sqlDB, dir, cleanup := backupRestoreTestSetupWithParams(t, MultiNode, numAccounts, InitManualReplication, params)
 	conn := sqlDB.DB.(*gosql.DB)
@@ -1328,9 +1318,55 @@ func checkInProgressBackupRestore(
 
 	sqlDB.Exec(t, `CREATE DATABASE restoredb`)
 
+	var totalExpectedBackupRequests int
+	// mergedRangeQuery calculates the number of spans we expect PartitionSpans to
+	// produce. It merges contiguous ranges on the same node.
+	// It sorts ranges by start_key and counts the number of times the
+	// lease_holder changes by comparing against the previous row's lease_holder.
+	mergedRangeQuery := `
+WITH
+	ranges
+		AS (
+			SELECT
+				start_key,
+				lag(lease_holder) OVER (ORDER BY start_key)
+					AS prev_lease_holder,
+				lease_holder
+			FROM
+				[SHOW RANGES FROM TABLE data.bank]
+		)
+SELECT
+	count(*)
+FROM
+	ranges
+WHERE
+	lease_holder != prev_lease_holder
+	OR prev_lease_holder IS NULL;
+`
+
+	sqlDB.QueryRow(t, mergedRangeQuery).Scan(&totalExpectedBackupRequests)
+
 	backupTableID := sqlutils.QueryTableID(t, conn, "data", "public", "bank")
 
 	do := func(query string, check inProgressChecker) {
+		t.Logf("checking query %q", query)
+
+		var totalExpectedResponses int
+		if strings.Contains(query, "BACKUP") {
+			// totalExpectedBackupRequests takes into account the merging that backup
+			// does of co-located ranges. It is the expected number of ExportRequests
+			// backup issues. DistSender will still split those requests to different
+			// ranges on the same node. Each range will write a file, so the number of
+			// SST files this backup will write is `backupRestoreDefaultRanges` .
+			totalExpectedResponses = totalExpectedBackupRequests
+		} else if strings.Contains(query, "RESTORE") {
+			// We expect restore to process each file in the backup individually.
+			// SST files are written per-range in the backup. So we expect the
+			// restore to process #(ranges) that made up the original table.
+			totalExpectedResponses = backupRestoreDefaultRanges
+		} else {
+			t.Fatal("expected query to be either a backup or restore")
+		}
 		jobDone := make(chan error)
 		allowResponse = make(chan struct{}, totalExpectedResponses)
 
@@ -1344,7 +1380,7 @@ func checkInProgressBackupRestore(
 			allowResponse <- struct{}{}
 		}
 
-		err := retry.ForDuration(testutils.DefaultSucceedsSoonDuration, func() error {
+		testutils.SucceedsSoon(t, func() error {
 			return check(ctx, inProgressState{
 				DB:            conn,
 				backupTableID: backupTableID,
@@ -1361,10 +1397,6 @@ func checkInProgressBackupRestore(
 		if err := <-jobDone; err != nil {
 			t.Fatalf("%q: %+v", query, err)
 		}
-
-		if err != nil {
-			t.Fatal(err)
-		}
 	}
 
 	do(`BACKUP DATABASE data TO $1`, checkBackup)
@@ -1373,7 +1405,6 @@ func checkInProgressBackupRestore(
 
 func TestBackupRestoreSystemJobsProgress(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 57831, "flaky test")
 	defer log.Scope(t).Close(t)
 	defer jobs.TestingSetProgressThresholds()()
 

--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -923,6 +923,10 @@ type BackupRestoreTestingKnobs struct {
 	// RunAfterProcessingRestoreSpanEntry allows blocking the RESTORE job after a
 	// single RestoreSpanEntry has been processed and added to the SSTBatcher.
 	RunAfterProcessingRestoreSpanEntry func(ctx context.Context)
+
+	// RunAfterExportingSpanEntry allows blocking the BACKUP job after a single
+	// span has been exported.
+	RunAfterExportingSpanEntry func(ctx context.Context)
 }
 
 var _ base.ModuleTestingKnobs = &BackupRestoreTestingKnobs{}

